### ISSUE TITLE
streamlit cloud 배포 후 Anomalib 감지되지 않음

### ISSUE THREAD

**Issue by HyunchanAn** [HyunchanAn/Surface-anomaly-detection#4]
<img width="1114" height="970" alt="Image" src="https://github.com/user-attachments/assets/bfa9ff8d-6c5f-4247-9b74-0b34195f55ca" />

최초 추정 원인: app.py에서 TorchInferencer와 OpenVINOInferencer를 하나의 try/except 블록으로 묶어서 임포트했는데, OpenVINOInferencer가 내부적으로 openvino 패키지를 로드하다가 Streamlit Cloud에서 실패하면서 TorchInferencer까지 통째로 임포트가 차단됨 

수정: 세 개의 임포트를 각각 독립적인 try/except로 분리

TorchInferencer → 단독 임포트 (이게 실패하면 ANOMALIB_AVAILABLE = False)
OpenVINOInferencer → 선택적 임포트 (없어도 앱 동작)
read_image → 선택적 임포트 (없어도 앱 동작)

> 해결되지 않음. 디버깅 메시지 출력

"Import error: No module named 'skimage'"
> scikit-image 패키지가 requirements.txt에 명시되지 않아 의존성 해석이 바뀐 것으로 추정되어 추가함 

"Import error: No module named 'lightning'"
> lightning 라이브러리 누락. requirements.txt에 추가

"Import error: No module named 'openvino'"
> TorchInferencer만 임포트해도 anomalib.deploy.__init__.py까지 통째로 로드되면서 openvino까지 불러오는 것으로 추정. __init__.py 우회하여 직접 파일에서 임포트할 예정

**Comment by HyunchanAn** [comment]
- 이슈 1 (altair 충돌): streamlit 버전 미명시로 구버전(1.19.0) 설치 → requirements.txt에 streamlit>=1.40.0 명시.
- 이슈 2 (CUDA 드라이버 없음): Streamlit Cloud는 CPU 환경이나 CUDA 포함 PyTorch 설치로 런타임 에러 발생 혹은 CPU 전용 PyTorch 설치 후 torch.cuda.is_available() 호출 시 AssertionError 발생.
    - app.py device 감지 로직에 try/except(Exception)을 적용하여, GPU 관련 어떤 에러(RuntimeError, AssertionError 등)가 발생해도 안전하게 CPU로 자동 전환되도록 수정.
    - requirements.txt에 --extra-index-url https://download.pytorch.org/whl/cpu 추가.
- 이슈 3 (anomalib import 실패 - openvino): OpenVINOInferencer와 TorchInferencer를 하나의 try/except로 묶은 구조에서, OpenVINOInferencer 임포트 시 openvino 패키지 부재로 예외 발생.
    - 1차 해결: TorchInferencer / OpenVINOInferencer / read_image 임포트를 각각 독립적인 try/except로 분리. 그리고 ANOMALIB_ERROR 변수를 도입하여 실제 에러 메시지를 사이드바에 표시.
    - 2차 해결: anomalib.deploy에서 TorchInferencer 임포트 시 내부 __init__.py를 거치며 다시 openvino를 로드하려는 문제 발견. 이를 우회하기 위해 anomalib.deploy.inferencers.torch_inferencer에서 직접 TorchInferencer를 임포트하도록 경로 수정.
- 이슈 4 (의존성 누락 - skimage, lightning, openvino): CPU 전용 PyTorch 인덱스 추가 후 uv의 의존성 해석이 변경되어 anomalib의 핵심 및 선택 의존성들이 누락.
    - 'No module named skimage' → scikit-image 추가.
    - 'No module named lightning' → lightning, pytorch-lightning, timm, einops, kornia, freia, scikit-learn 일괄 추가.
    - 'No module named openvino' → 임포트 우회 시도에도 에러 발생하여, 아예 openvino 패키지를 명시적으로 추가하고 anomalib==2.2.0으로 버전을 고정하여 근본적 해결.
- 이슈 5 (torchmetrics GPU 속성 유지 에러): GPU에서 학습된 모델을 CPU 전용 환경에서 로드하여 추론(device="cpu")할 때, 모델 내부에 저장된 torchmetrics 객체들이 여전히 자신의 장치를 "cuda:0"으로 기억하고 있어, dummy tensor 생성 시 AssertionError("Torch not compiled with CUDA enabled") 발생.
    - 해결: app.py 상단에 torchmetrics.Metric._apply 메서드를 몽키 패치(Monkey Patching)하여, torch.cuda.is_available()이 False일 경우 metric 객체의 내부 _device 속성이 "cuda"를 포함하면 강제로 "cpu"로 변경하도록 조치.
- 배포 URL: sg-proj-005-sad.streamlit.app
